modules/mauth-authenticator-http4s/src/main/scala/com/mdsol/mauth/http4s/RequestAuthenticator.scala

```diff
@@ -0,0 +1,155 @@
+package com.mdsol.mauth.http4s
+
+import cats.effect.Async
+import cats.implicits._
+import com.mdsol.mauth.{MAuthRequest, MAuthVersion}
+import com.mdsol.mauth.exception.MAuthValidationException
+import com.mdsol.mauth.http4s.RequestAuthenticator.{fallbackValidateSignatureV1, validateMauthVersion, validateSignatureV1, validateSignatureV2, validateTime}
+import com.mdsol.mauth.scaladsl.Authenticator
+import com.mdsol.mauth.scaladsl.utils.ClientPublicKeyProvider
+import com.mdsol.mauth.util.{EpochTimeProvider, MAuthSignatureHelper}
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import java.nio.charset.StandardCharsets
+import java.security.PublicKey
+import java.util
+import scala.concurrent.duration._
+
+class RequestAuthenticator[F[_]: Async](
+  val publicKeyProvider: ClientPublicKeyProvider[F],
+  override val epochTimeProvider: EpochTimeProvider,
+  v2OnlyAuthenticate: Boolean
+) extends Authenticator[F] {
+
+  implicit val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLoggerFromClass(classOf[RequestAuthenticator[F]])
+
+  /** check if mauth v2 only authenticate is enabled or not
+    *
+    * @return True or false identifying if v2 only authenticate is enabled or not.
+    */
+  override val isV2OnlyAuthenticate: Boolean = v2OnlyAuthenticate
+
+  /** Performs the validation of an incoming HTTP request.
+    *
+    * The validation process consists of recreating the mAuth hashed signature from the request data
+    * and comparing it to the decrypted hash signature from the mAuth header.
+    *
+    * @param mAuthRequest Data from the incoming HTTP request necessary to perform the validation.
+    * @return True or false indicating if the request is valid or not with respect to mAuth.
+    */
+  override def authenticate(mAuthRequest: MAuthRequest)(implicit requestValidationTimeout: Duration): F[Boolean] = {
+    if (!validateTime(mAuthRequest.getRequestTime, epochTimeProvider)(requestValidationTimeout)) {
+      val message = s"MAuth request validation failed because of timeout $requestValidationTimeout"
+      logger.error(message) *> Async[F].raiseError(new MAuthValidationException(message))
+    } else if (!validateMauthVersion(mAuthRequest, v2OnlyAuthenticate)) {
+      val message = "The service requires mAuth v2 authentication headers."
+      logger.error(message) *> Async[F].raiseError(new MAuthValidationException(message))
+    } else {
+      getPublicKey(mAuthRequest)
+    }
+  }
+
+  private def getPublicKey(mAuthRequest: MAuthRequest): F[Boolean] = {
+    publicKeyProvider.getPublicKey(mAuthRequest.getAppUUID).map {
+      case None =>
+        logger.error("Public Key couldn't be retrieved")
+        false
```

Comment:
this wouldn't work - when you operate in some `F` context that gets evaluated by the cats effect runtime, all operations need to be chained. So typically this would look like
`logger.error(...) *> false.pure`